core/src/main/java/org/apache/accumulo/fate/zookeeper/DistributedReadWriteLock.java

```diff
@@ -218,22 +237,54 @@ public boolean tryLock() {
       }
       SortedMap<Long,byte[]> entries = qlock.getEarlierEntries(entry);
       Iterator<Entry<Long,byte[]>> iterator = entries.entrySet().iterator();
-      if (!iterator.hasNext())
+      if (!iterator.hasNext()) {
         throw new IllegalStateException("Did not find our own lock in the queue: " + this.entry
             + " userData " + new String(this.userData, UTF_8) + " lockType " + lockType());
-      return iterator.next().getKey().equals(entry);
+      }
+      if (!failBlockers) {
+        return iterator.next().getKey().equals(entry);
+      } else {
+        ZooStore<DistributedReadWriteLock> zs;
+        try {
+          zs = new ZooStore<>(zooPath, zrw);
+        } catch (KeeperException | InterruptedException e1) {
+          log.error("Error creating zoo store", e1);
+          return false;
+        }
+        final AdminUtil<DistributedReadWriteLock> util = new AdminUtil<>();
+        boolean result = true;
+        while (iterator.hasNext()) {
+          Entry<Long,byte[]> e = iterator.next();
+          if (!e.getKey().equals(entry)) {
+            result &= util.prepFail(zs, zrw, zooManagerPath, Long.toString(e.getKey(), 16));
```

Comment:
I see two problems with the call here.  First prepFail() calls  checkGlobalLock() which calls System.exit() when the constuctor AdminUtil<>() is called.  The reason it calles checkGlobalLock is because I think this code was only intended to be called when the manager was not running and nothing was running.  The Second problem is that prepFail only initiates failure, it does not wait for any operations that may currently be actively executing and holding the lock to finish.  This is why prepFail checks that the Manager is not running, because it does not deal with this concurrency issue any way.  So maybe something could get the lock by failing other operations, but those failed operation could still be running and mutating the metadata table and zookeeper after the lock is acquired.

The first problem is easy to address by calling a diff constructor, but the 2nd problem would require some new mechanism to initiate failure and wait for anything that may be running. Since all fate ops are run in the manager could have something that does the following.

- Prevents repos related to failed txids from starting to run (like if it was sitting a thread pool queue)
- Waits until repos related to failed txids that are currently running complete.

Reply:
Looking at this some more, the call to reserve() in prepFail may address the 2nd issue I mentioned above

Reply:
I think as long as the manager and this code are using the same zoostore object, then the call to reserve will address the 2nd problem I mentioned above.  I see a third problem, prepFail transitions FATE ops to FAILED_IN_PROGRESS.  This means those fate ops will unwind an execute their undo() operations which could modify the metadata table and zookeeper. So this could lead to other FATE ops that no longer hold the lock still modifying persisted state related to the table.  Would probably need to wait for these FATE ops to transition from FAILED_IN_PROGRESS to FAILED before getting the lock.   Have to be careful about how this wait is done.  If all threads in the fate thread pool are waiting for other operations to transition from  FAILED_IN_PROGRESS to FAILED, then that would not leave any threads available to transitions those operations leading to deadlock.

Reply:
A fourth issue to consider is deadlock around the reserve() calls, which I think may not happen but I am not completely sure.  Calls to tryLock are made by FATE ops that have already reserved a txid.  Assume FATE op txidA is reserved and it runs and calls tryLock which attempts to reserve tixdB. If txidB is reserved and it calls tryLock trying to reserved txidA that leads to deadlock.  I think this change to tryLock will only attempt to reserve transactions that started earlier than itself which seems like it would avoid deadlock.  Because if A started before B, then B would attempt to reserve A but A would never attempt to reserve B because it started after.

Reply:
@keith-turner  - you probably know this code best, can you suggest an alternate implementation?

Reply:
@dlmarion  could possible do something like the following... where the code in  `failPrecedingOperations()` would go through and transition operations to FAILED_IN_PROGRESS... also thinking that code could transition NEW things straight to FAILED instead of to FAILED_IN_PROGRESS but not sure.   However there is not waiting on these other operations.   The FATE framework will reserve a txid, take the top repo off the stack, and then call isReady().  If isReady returns non-zero it will unreserve and the txid and come back to it later.  So if isReady transitions things to FAILED_IN_PROGRESS then hopefully a later call to isReady can get the lock sooner.

```java
   class SomeRepo extends Repo {
        public long isReady(long tid,...) {
            long waitTime = tryToGetTableWriteLock(...);

           if(waitTime > 0 && failBlockers) {
                // we did not get the the table write lock, so lets initiate the failure of all operations that precede us in the lock queue.... hopefully after these fail they will release the lock and we will get the lock on a later call to isRead()
               failPrecedingOperations(tid)
           }
           
           return waitTime;
        }
    
   }
```

Reply:
For the prev comment, failPrecedingOperations() code could possibly be pushed into the lock code.  I teased it out in the example above to show how it could work w/ the isReady() call to avoid waiting for transitions to happen.